include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_utils.h

```diff
@@ -0,0 +1,881 @@
+    __block_store<_T, _N - __block_size>(__dst + __block_size, __data.template select<_N - __block_size, 1>(__block_size));
+}
+
+template <typename _T, int _N>
+inline std::enable_if_t<(_N > 16) && (_N % 16 == 0), __dpl_esimd_ns::simd<_T, _N>>
+__create_simd(_T initial, _T step)
+{
+    __dpl_esimd_ns::simd<_T, _N> ret;
+    ret.template select<16, 1>(0) = __dpl_esimd_ns::simd<_T, 16>(0, 1) * step + initial;
+    __dpl_esimd_ns::fence<__dpl_esimd_ns::fence_mask::sw_barrier>();
+    _ONEDPL_PRAGMA_UNROLL
+    for (int pos = 16; pos < _N; pos += 16)
+    {
+        ret.template select<16, 1>(pos) = ret.template select<16, 1>(0) + pos * step;
+    }
+    return ret;
+}
+
+template <typename _T>
+struct __slm_lookup
+{
+    ::std::uint32_t __slm;
+    inline __slm_lookup(::std::uint32_t __slm) : __slm(__slm) {}
+
+    template <int __table_size>
+    inline void
+    __setup(__dpl_esimd_ns::simd<_T, __table_size> __source) SYCL_ESIMD_FUNCTION
+    {
+        __utils::__block_store_slm<_T, __table_size>(__slm, __source);
+    }
+
+    template <int _N, typename _Idx>
+    inline auto
+    __lookup(_Idx __idx) SYCL_ESIMD_FUNCTION
+    {
+        return __utils::__vector_load<_T, 1, _N>(__slm + __dpl_esimd_ns::simd<::std::uint32_t, _N>(__idx) * sizeof(_T));
+    }
+
+    template <int _N, int __table_size, typename _Idx>
+    inline auto
+    __lookup(__dpl_esimd_ns::simd<_T, __table_size> __source, _Idx __idx) SYCL_ESIMD_FUNCTION
+    {
+        __setup(__source);
+        return __lookup<_N>(__idx);
+    }
+};
+
+template<typename _Rng>
+auto
+__rng_data(const _Rng& __rng)
+{
+    return __rng.begin();
+}
+
+// ESIMD functionality requires using an accessor directly due to the restriction:
+//      sycl::accessor::operator[] are supported only with -fsycl-esimd-force-stateless-mem.
+//      Otherwise, all memory accesses through an accessor are done via explicit APIs
+// TODO: rely on begin() once -fsycl-esimd-force-stateless-mem has been enabled by default
+template <typename _T, sycl::access::mode _M>
+auto
+__rng_data(const oneapi::dpl::__ranges::all_view<_T, _M>& __view)
+{
+    return __view.accessor();
+}
+
+struct __rng_dummy {};
+
+template<typename _Rng>
+struct __rng_value_type_deducer
+{
+    using __value_t = oneapi::dpl::__internal::__value_t<_Rng>;
+};
+
+template<>
+struct __rng_value_type_deducer<__rng_dummy>
+{
+    using __value_t = void;
+};
+
+template<typename _Rng1, typename _Rng2 = __rng_dummy>
+struct __rng_pack
+{
+    using _KeyT = typename __rng_value_type_deducer<_Rng1>::__value_t;
+    using _ValT = typename __rng_value_type_deducer<_Rng2>::__value_t;
+    static constexpr bool __has_values = !std::is_void_v<_ValT>;
+
+    const auto& __keys_rng() const { return __m_keys_rng; }
+    const auto& __vals_rng() const
+    {
+        static_assert(__has_values);
+        return __m_vals_rng;
+    }
+
+    __rng_pack(const _Rng1& __rng1, const _Rng2& __rng2 = __rng_dummy{}):
```

Comment:
Looks like these two constructors may be combined into one with universal reference

Reply:
This will become obsolete if we remove `__rng_pack` entirely, e.g. if we use `zip_iterator` or `zip_view`. But thank you for the suggestion.